rust/src/applayertemplate/mod.rs

```diff
@@ -0,0 +1,518 @@
+                index += 1;
+                continue;
+            }
+            *state = index as u64 + 1;
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
+        }
+
+        return None;
+    }
+}
+
+/// Probe to see if this input looks like a request or response.
+///
+/// For the purposes of this template things will be kept simple. The
+/// protocol is text based with the leading text being the length of
+/// the message in bytes. So simply make sure the first character is
+/// between "1" and "9".
+fn probe(input: &[u8]) -> bool {
+    if input.len() > 1 && input[0] >= 49 && input[0] <= 57 {
+        return true;
+    }
+    return false;
+}
+
+// C exports.
+
+export_tx_get_detect_state!(
+    rs_template_tx_get_detect_state,
+    TemplateTransaction
+);
+export_tx_set_detect_state!(
+    rs_template_tx_set_detect_state,
+    TemplateTransaction
+);
+
+/// C entry point for a probing parser.
+#[no_mangle]
+pub extern "C" fn rs_template_probing_parser(
+    _flow: *const Flow,
+    input: *const libc::uint8_t,
+    input_len: u32,
+) -> AppProto {
+    // Need at least 2 bytes.
+    if input_len > 1 && input != std::ptr::null_mut() {
+        let slice = build_slice!(input, input_len as usize);
+        if probe(slice) {
+            return unsafe { ALPROTO_TEMPLATE };
+        }
+    }
+    return ALPROTO_UNKNOWN;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_template_state_new() -> *mut libc::c_void {
+    let state = TemplateState::new();
+    let boxed = Box::new(state);
+    return unsafe { transmute(boxed) };
+}
+
+#[no_mangle]
+pub extern "C" fn rs_template_state_free(state: *mut libc::c_void) {
+    // Just unbox...
+    let _drop: Box<TemplateState> = unsafe { transmute(state) };
+}
+
+#[no_mangle]
+pub extern "C" fn rs_template_state_tx_free(
+    state: *mut libc::c_void,
+    tx_id: libc::uint64_t,
+) {
+    let state = cast_pointer!(state, TemplateState);
+    state.free_tx(tx_id);
+}
+
+#[no_mangle]
+pub extern "C" fn rs_template_parse_request(
+    _flow: *const Flow,
+    state: *mut libc::c_void,
+    _pstate: *mut libc::c_void,
+    input: *const libc::uint8_t,
+    input_len: u32,
+    _data: *const libc::c_void,
+    _flags: u8,
+) -> i8 {
+    let _eof = unsafe {
+        if AppLayerParserStateIssetFlag(_pstate, APP_LAYER_PARSER_EOF) > 0 {
+            true
+        } else {
+            false
+        }
+    };
+    let state = cast_pointer!(state, TemplateState);
+    let buf = build_slice!(input, input_len as usize);
+    if state.parse_request(buf) {
+        return 1;
+    }
+    return -1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_template_parse_response(
+    _flow: *const Flow,
+    state: *mut libc::c_void,
+    _pstate: *mut libc::c_void,
+    input: *const libc::uint8_t,
+    input_len: u32,
+    _data: *const libc::c_void,
+    _flags: u8,
+) -> i8 {
+    let _eof = unsafe {
+        if AppLayerParserStateIssetFlag(_pstate, APP_LAYER_PARSER_EOF) > 0 {
```

Comment:
_pstate is a bit ugly here, the leading _ should mean it's not used?

Reply:
Oops. Renamed to pstate. Also using eof, but only like
```
diff --git a/rust/src/applayertemplate/mod.rs b/rust/src/applayertemplate/mod.rs
index 9daa161b5..5bb6705a5 100644
--- a/rust/src/applayertemplate/mod.rs
+++ b/rust/src/applayertemplate/mod.rs
@@ -300,19 +300,24 @@ pub extern "C" fn rs_template_state_tx_free(
 pub extern "C" fn rs_template_parse_request(
     _flow: *const Flow,
     state: *mut libc::c_void,
-    _pstate: *mut libc::c_void,
+    pstate: *mut libc::c_void,
     input: *const libc::uint8_t,
     input_len: u32,
     _data: *const libc::c_void,
     _flags: u8,
 ) -> i8 {
-    let _eof = unsafe {
-        if AppLayerParserStateIssetFlag(_pstate, APP_LAYER_PARSER_EOF) > 0 {
+    let eof = unsafe {
+        if AppLayerParserStateIssetFlag(pstate, APP_LAYER_PARSER_EOF) > 0 {
             true
         } else {
             false
         }
     };
+
+    if eof {
+        // If needed, handled EOF, or pass it into the parser.
+    }
+
     let state = cast_pointer!(state, TemplateState);
     let buf = build_slice!(input, input_len as usize);
     if state.parse_request(buf) {
```

The idea with the template is to just show how to test for EOF. Do you think we should do anything else with it here?

Reply:
Not at this time I think. It's more used to close things like files.